src/uct/ib/ud/base/ud_ep.c

```diff
@@ -288,7 +288,15 @@ static void uct_ud_ep_timer_backoff(uct_ud_ep_t *ep)
 
 static UCS_F_ALWAYS_INLINE int uct_ud_ep_is_last_ack_received(uct_ud_ep_t *ep)
 {
-    return UCT_UD_PSN_COMPARE(ep->tx.acked_psn, ==, ep->tx.psn - 1);
+    int ack_recvd = UCT_UD_PSN_COMPARE(ep->tx.acked_psn, ==, ep->tx.psn - 1);
+
+    if (!ack_recvd) {
+        ucs_assertv(!ucs_queue_is_empty(&ep->tx.window),
+                    "ep %p: acked_psn=%u current_psn=%u", ep, ep->tx.acked_psn,
+                    ep->tx.psn);
+    }
```

Comment:
better move the assert outside of uct_ud_ep_is_last_ack_received() since this function should just check a condition on PSN and is not directly related to window size. A function should be well-defined.

If the goal is to not duplicate the printing of PSNs, we can add new wrapper uct_ud_ep_assert_tx_window_nonempty(ep)

Reply:
thanks, good idea
added `uct_ud_ep_assert_tx_window_nonempty`